compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -480,3 +539,75 @@ splitUnitId unitId = fromMaybe (unitId, Nothing) $ do
     (name, ver) <- stripInfixEnd "-" unitId
     guard $ all (`elem` '.' : ['0' .. '9']) ver
     pure (name, Just ver)
+
+-- | Returns 'True' if an LF type contains a reference to an
+-- old-style typeclass. See 'tconIsOldTypeclass' for more details.
+typeHasOldTypeclass :: Env -> LF.Type -> Bool
+typeHasOldTypeclass env = \case
+    LF.TVar _ -> False
+    LF.TCon tcon -> tconIsOldTypeclass env tcon
+    LF.TSynApp _ _ -> True
+        -- Type synonyms came after the switch to new-style
+        -- typeclasses, so we can assume there are no old-style
+        -- typeclasses being referenced. HOWEVER, we don't support
+        -- type synonyms here yet. TODO: Fix this, and change
+        -- the above to False.
+    LF.TApp a b -> typeHasOldTypeclass env a || typeHasOldTypeclass env b
+    LF.TBuiltin _ -> False
+    LF.TForall _ b -> typeHasOldTypeclass env b
+    LF.TStruct fields -> any (typeHasOldTypeclass env . snd) fields
+    LF.TNat _ -> False
+
+-- | Determine whether a typecon refers to an old-style
+-- typeclass. By "old-style" I mean a typeclass based on
+-- nominal LF record types. There's no foolproof way of
+-- determining this, but we can approximate it by taking
+-- advantage of the typeclass sanitization that is introduced
+-- during LF conversion: every field in a typeclass dictionary
+-- is represented by a type @() -> _@ due to sanitization.
+-- Thus, if a type is given by a record, and the record has
+-- at least one field, and every field in the record is a
+-- function from unit, then there's a good chance this
+-- record represents an old-style typeclass.
+--
+-- The caveat here is that if a user creates a record that
+-- matches these criteria, it will be treated as an old-style
+-- typeclass and therefore any functions that use this type
+-- will not be exposed via the data-dependency mechanism,
+-- (see 'generateSrcFromLf').
+tconIsOldTypeclass :: Env -> LF.Qualified LF.TypeConName -> Bool
+tconIsOldTypeclass env tcon =
+    case envLookupDataType tcon env of
+        Nothing -> error ("Unknown reference to type " <> show tcon)
+        Just dtype -> defDataTypeIsOldTypeClass dtype
+
+defDataTypeIsOldTypeClass :: LF.DefDataType -> Bool
+defDataTypeIsOldTypeClass LF.DefDataType{..}
+    | LF.DataRecord fields <- dataCons
+    = notNull fields && all isDesugarField fields
```

Comment:
Fun fact: We have typeclasses with no methods in `daml-stdlib` which this will not hit but it should be fine to not handle that and I don’t see a reasonable way of detecting those.